File: Estructura de Datos/Relacion de ejercicios/Examenes/2.-Septiembre 2020/Examen Septiembre 2020-20250109/haskell/haskell/AVL.hs
{------------------------------------------------------------------------------
 - Student's name:
 -
 - Student's group:
 -----------------------------------------------------------------------------}

module AVL 
  ( 
    Weight
  , Capacity
  , AVL (..)
  , Bin
  , emptyBin
  , remainingCapacity
  , addObject
  , maxRemainingCapacity
  , height
  , nodeWithHeight
  , node
  , rotateLeft
  , addNewBin
  , addFirst
  , addAll
  , toList
  , linearBinPacking
  , seqToList
  , addAllFold
  ) where

type Capacity = Int
type Weight= Int

data Bin = B Capacity [Weight] 

data AVL = Empty | Node Bin Int Capacity AVL AVL deriving Show

--) (0,5 puntos) Define las siguientes funciones auxiliares: 

--A- emptyBin que dada una capacidad devuelve un cubo vacío con tal capacidad 
emptyBin :: Capacity -> Bin
emptyBin capacity = B capacity []

--B- remainingCapacity que dado un cubo devuelve su capacidad restante 

remainingCapacity :: Bin -> Capacity
remainingCapacity (B capacidad weigh) = capacidad - sum weigh

--C- addObject que dados un cubo y un objeto lo añade al cubo; si el objeto no cabe en el cubo 
--señala un error 
addObject :: Weight -> Bin -> Bin
addObject obj (B capacidad weigh) = 
    if obj <= (capacidad - sum weigh) then (B capacidad (obj:weigh))
    else error "El objeto no cabe en el cubo"


--D maxRemainingCapacity que dado un AVL devuelva su capacidad restante máxima 

maxRemainingCapacity :: AVL -> Capacity
maxRemainingCapacity Empty = 0
maxRemainingCapacity (Node bin _ _ left right)
    = max (remainingCapacity bin) (max(maxRemainingCapacity left) (maxRemainingCapacity right))

--E- height que dado un AVL devuelve su altura

height :: AVL -> Int
height Empty = 0
height (Node bin heigh _ _ _) = heigh 

--b) (0,5 puntos) Define una función nodeWithHeight que tome un cubo, una altura, un AVL (hijo 
--izquierdo) y otro AVL (hijo derecho) y que devuelva un AVL con dicha información, calculando su 
--capacidad restante máxima.
 
nodeWithHeight :: Bin -> Int -> AVL -> AVL -> AVL
nodeWithHeight bin h left right =
  let maxCap = maximum [remainingCapacity bin, maxRemainingCapacity left, maxRemainingCapacity right]
  in Node bin h maxCap left right

--c) (0,5 puntos) Define una función node que tome un cubo, un AVL (hijo izquierdo) y otro AVL (hijo 
---derecho) y que devuelva un AVL con dicha información, calculando su capacidad restante máxima 
--y la correspondiente altura.

node :: Bin -> AVL -> AVL -> AVL
node bin left right = 
  let h= 1+max (height left) (height right)
  in nodeWithHeight bin h left right

--d) (1 punto) Define una función rotateLeft que tome un cubo c, un AVL (hijo izquierdo) L y otro 
--AVL no vacío (hijo derecho) R y que devuelva un AVL creado con dicha información tras aplicar, 
--además, una rotación simple a la izquierda tal como se ve en la siguiente figura (dado que el árbol R 
--no está vacío, se muestran su raíz, x, y sus hijos R1 y R2):

rotateLeft :: Bin -> AVL -> AVL -> AVL
rotateLeft c left (Node x hx mx leftR rightR)=
  let newLeft = node c left leftR
  in node x newLeft rightR
rotatreLeft _ _ Empty = error "El árbol derecho no puede ser vacío"

--e) (1,5 puntos) Define una función addNewBin que tome un cubo y un árbol AVL y que añada un 
--nuevo nodo con dicho cubo al final de la espina derecha del AVL. Para mantener el invariante de los 
--árboles AVL, en cada nodo de la espina derecha, si la altura resultante del hijo derecho es más de una 
--unidad superior a la del hijo izquierdo, habrá que aplicar una rotación simple a la izquierda. 

addNewBin :: Bin -> AVL -> AVL
addNewBin bin Empty = node bin Empty Empty
addNewBin bin (Node root h maxcap left right)=
  let newRight = addNewBin bin right
  in if height newRight > height left + 1
    then rotateLeft root left newRight
    else node root left newRight
 
--f) (2.0 puntos) Define una función addFirst que tome la capacidad de los cubos del problema (W), 
--el peso de un objeto a añadir y un AVL y que añada dicho objeto al primer cubo que pueda albergarlo 
--o añada un nuevo cubo al final de la espina derecha si el nuevo objeto no cabe en ningún cubo. El 
--algoritmo será el siguiente: 
-- Si el AVL está vacío o no cabe en ningún cubo, se añadirá un nuevo nodo con un cubo con el 
--objeto al final de la espina derecha. 
-- En otro caso, si la capacidad restante máxima del hijo izquierdo es mayor o igual al peso del 
--objeto, se añadirá el objeto al primer cubo posible del hijo izquierdo. 
-- En otro caso, si la capacidad restante del cubo en el nodo raíz es mayor o igual al peso del 
--objeto, se añadirá el objeto al cubo en la raíz. 
-- En otro caso, se añadirá el objeto al primer cubo posible del hijo derecho. 

addFirst :: Capacity -> Weight -> AVL -> AVL
addFirst w obj Empty = addNewBin (addObject obj (emptyBin w)) Empty
addFirst w obj (Node bin h maxcap left right) 
  | obj <= maxRemainingCapacity left = node bin (addFirst w obj left) right
  | obj <= remainingCapacity bin = node (addObject obj bin) left right
  | otherwise = node bin left (addFirst w obj right)

--g) (0,75 puntos) Define una función addAll que tome el valor de la capacidad máxima de los cubos 
--(W), una lista de pesos de objetos y que construya un AVL con cubos que contenga todos los objetos 
--de la lista, según se ha descrito anteriormente. 

addAll:: Capacity -> [Weight] -> AVL
addAll w = foldl ( flip (addFirst w)) Empty

--h) (0,75 puntos) Define una función toList que tome un AVL y devuelva una lista de cubos con su 
--recorrido en en-orden.

toList :: AVL -> [Bin]
toList Empty = []
toList (Node bin _ _ left right) =
  toList left ++ [bin] ++ toList right

{-
	SOLO PARA ALUMNOS SIN EVALUACION CONTINUA
  ONLY FOR STUDENTS WITHOUT CONTINUOUS ASSESSMENT
 -}

data Sequence = SEmpty | SNode Bin Sequence deriving Show  

linearBinPacking:: Capacity -> [Weight] -> Sequence
linearBinPacking _ _ = undefined

seqToList:: Sequence -> [Bin]
seqToList _ = undefined

addAllFold:: [Weight] -> Capacity -> AVL 
addAllFold _ _ = undefined



{- No modificar. Do not edit -}

objects :: Bin -> [Weight]
objects (B _ os) = reverse os

  
instance Show Bin where
  show b@(B c os) = "Bin("++show c++","++show (objects b)++")"

File: Estructura de Datos/Relacion de ejercicios/Examenes/2.-Septiembre 2020/Examen Septiembre 2020-20250109/java/java/src/AVL.java
/**
 * Student's name:
 *
 * Student's group:
 */

import dataStructures.list.List;
import dataStructures.list.ArrayList;
import dataStructures.list.LinkedList;

import java.util.Iterator;


class Bin {
    private int remainingCapacity; // capacity left for this bin
    private List<Integer> weights; // weights of objects included in this bin

    public Bin(int initialCapacity) {
        this.remainingCapacity = initialCapacity;
        this.weights = new ArrayList<>();
    }

    // returns capacity left for this bin
    public int remainingCapacity() {
        return remainingCapacity;
    }

    // adds a new object to this bin
    public void addObject(int weight) {
        if(weight > remainingCapacity){
            throw new IllegalArgumentException("Weight exceeds remaining capacity");
        }
        else {
            remainingCapacity -= weight;
            weights.append(weight);
        }
    }

    // returns an iterable through weights of objects included in this bin
    public Iterable<Integer> objects() {
        
        return weights;
    }

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        sb.append(remainingCapacity);
        sb.append(", ");
        sb.append(weights.toString());
        sb.append(")");
        return sb.toString();
    }
}

// Class for representing an AVL tree of bins
public class AVL {
    static private class Node {
        Bin bin; // Bin stored in this node
        int height; // height of this node in AVL tree
        int maxRemainingCapacity; // max capacity left among all bins in tree rooted at this node
        Node left, right; // left and right children of this node in AVL tree

        Node(Bin bin){
            this.bin = bin;
            this.height = 1;
            this.maxRemainingCapacity = bin.remainingCapacity();
            this.left = null;
            this.right = null;
        }
        // recomputes height of this node
        void setHeight() {
            this.height = 1 + Math.max(height(left), height(right));
        }

        // recomputes max capacity among bins in tree rooted at this node
        void setMaxRemainingCapacity() {
            this.maxRemainingCapacity = Math.max( bin.remainingCapacity(), Math.max(maxRemainingCapacity(left), maxRemainingCapacity(right)) );
        }

        // left-rotates this node. Returns root of resulting rotated tree
        Node rotateLeft() {
            Node newRoot = this.right;
            this.right = newRoot.left;
            newRoot.setHeight();
            newRoot.setMaxRemainingCapacity();
            return newRoot;
        }
    }

    private static int height(Node node) {
        int height = 0;
        if(node!=null){
            height =node.height;
        }
        return height;
    }

    private static int maxRemainingCapacity(Node node) {
        int remaincap = 0;

        if(node!=null){
            remaincap = node.maxRemainingCapacity;
        }
        return remaincap;
    }

    private Node root; // root of AVL tree

    public AVL() {
        this.root = null;
    }

    // adds a new bin at the end of right spine.
    private void addNewBin(Bin bin) {
        
        root = addNewBin(root,bin);
    }
    private Node addNewBin(Node n, Bin bin) {
        if (n == null) {
            Node nod= new Node(bin);
            return nod;
        }

        n.right = addNewBin(n.right, bin);

        if (height(n.right) > height(n.left) + 1) {
            n = n.rotateLeft();
        } else {
            n.setHeight();
            n.setMaxRemainingCapacity();
        }

        return n;
    }
    // adds an object to first suitable bin. Adds
    // a new bin if object cannot be inserted in any existing bin
/*f) (2.0 puntos) 
Define el método void addFirst(int initialCapacity, int weight) de 
la clase AVL que tome la capacidad de los cubos del problema (W), el peso de un objeto a añadir y 
que añada dicho objeto al primer cubo que pueda albergarlo o añada un nuevo cubo al final de la 
espina derecha si el nuevo objeto no cabe en ningún cubo. El algoritmo será el siguiente: 
 Si el AVL está vacío o no cabe en ningún cubo, se añadirá un nuevo nodo con un cubo con el 
objeto al final de la espina derecha. 
 En otro caso, si la capacidad restante máxima del hijo izquierdo es mayor o igual al peso del 
objeto, se añadirá el objeto al primer cubo posible del hijo izquierdo. 
 En otro caso, si la capacidad restante del cubo en el nodo raíz es mayor o igual al peso del 
objeto, se añadirá el objeto al cubo en la raíz. 
 En otro caso, se añadirá el objeto al primer cubo posible del hijo derecho.  */
    public void addFirst(int initialCapacity, int weight) {
        root = addFirst(root,initialCapacity,weight);
    }
    private Node addFirst(Node n, int initialCapacity, int weight) {
        if(n == null ){
            Bin newBin = new Bin(initialCapacity);
            newBin.addObject(weight);
            return new Node(newBin);
        }

        if(weight <= maxRemainingCapacity(n.left)){
            n.left = addFirst(n.left, initialCapacity, weight);
        }else if(weight <= n.bin.remainingCapacity()){
            n.bin.addObject(weight);
        }else{
            n.right = addFirst(n.right,initialCapacity,weight);
        }

        n.setHeight();
        n.setMaxRemainingCapacity();

        return n;
    }
  

    public void addAll(int initialCapacity, int[] weights) {
        for (int weight : weights) {
            addFirst(initialCapacity, weight);
        }
    }
    

    public List<Bin> toList() {
        List<Bin> binList = new ArrayList<>();
        toListHelper(root, binList);
        return binList;
    }
    
    private void toListHelper(Node node, List<Bin> binList) {
        if (node != null) {
            toListHelper(node.left, binList);
            binList.append(node.bin);
            toListHelper(node.right, binList);
        }
    }
    

    public String toString() {
        String className = getClass().getSimpleName();
        StringBuilder sb = new StringBuilder(className);
        sb.append("(");
        stringBuild(sb, root);
        sb.append(")");
        return sb.toString();
    }

    private static void stringBuild(StringBuilder sb, Node node) {
        if(node==null)
            sb.append("null");
        else {
            sb.append(node.getClass().getSimpleName());
            sb.append("(");
            sb.append(node.bin);
            sb.append(", ");
            sb.append(node.height);
            sb.append(", ");
            sb.append(node.maxRemainingCapacity);
            sb.append(", ");
            stringBuild(sb, node.left);
            sb.append(", ");
            stringBuild(sb, node.right);
            sb.append(")");
        }
    }
}

class LinearBinPacking {
    public static List<Bin> linearBinPacking(int initialCapacity, List<Integer> weights) {
        List<Bin> bins = new ArrayList<>();

        for(int weight : weights){
            boolean placed = false;
            for(Bin bin: bins){
                if(bin.remainingCapacity() >= weight){
                    bin.addObject(weight);
                    placed = true;
                    break;
                }
            }
            if(!placed){
                Bin newBin = new Bin (initialCapacity);
                newBin.addObject(weight);
            }
        }
        return bins;
    }
	
	public static Iterable<Integer> allWeights(Iterable<Bin> bins) {
        List<Integer> allweighs = new ArrayList<>();
        for(Bin bin: bins){
            for(int weight: bin.objects()){
                allweighs.append(weight);
            }
        }
        return allweighs;		
	}
}